### PR TITLE
Add docs for multi-stage builds

### PR DESCRIPTION
**Pull request by tonistiigi**
cc @thaJeztah 

Signed-off-by: Tonis Tiigi <tonistiigi@gmail.com>

### CODE REVIEW

File: docs/reference/builder.md [hunk 1]
```diff
@@ -480,30 +480,36 @@ the first pattern, followed by one or more `!` exception patterns.
...
-for subsequent instructions. As such, a valid `Dockerfile` must have `FROM` as
-its first instruction. The image can be any valid image – it is especially easy
-to start by **pulling an image** from the [*Public Repositories*](https://docs.docker.com/engine/tutorials/dockerrepos/).
+The `FROM` instruction initializes a new build stage and sets the 
+[*Base Image*](glossary.md#base-image) for subsequent instructions. As such, a 
+valid `Dockerfile` must have `FROM` as its first instruction. The image can be
+any valid image – it is especially easy to start by **pulling an image** from 
+the [*Public Repositories*](https://docs.docker.com/engine/tutorials/dockerrepos/).
 
 - `FROM` must be the first non-comment instruction in the `Dockerfile`.
```
Comment (AkihiroSuda) [inline docs/reference/builder.md]: This line seems outdated. Now we can have ARG.

Comment (tonistiigi) [inline docs/reference/builder.md]: not yet: https://github.com/docker/docker/pull/31352

**Comment by vieux** [comment]
SGTM ping @mstanleyjones

**Comment by vieux** [comment]
ping @thaJeztah

File: docs/reference/builder.md [hunk 2]
```diff
@@ -480,30 +480,36 @@ the first pattern, followed by one or more `!` exception patterns.
...
 
-- The `tag` or `digest` values are optional. If you omit either of them, the builder
-assumes a `latest` by default. The builder returns an error if it cannot match
-the `tag` value.
+- Optionally a name can be given to a new build stage. That name can be then
+used in subsequent `FROM` and `COPY --from=` commands to refer back to the 
+image built in this stage.
+
+- The `tag` or `digest` values are optional. If you omit either of them, the 
+builder assumes a `latest` by default. The builder returns an error if it
```
Comment (thaJeztah) [inline docs/reference/builder.md]: nit: perhaps be explicit;

`latest` _tag_ 

(as it's not latest "digest")

File: docs/reference/builder.md [hunk 3]
```diff
@@ -480,30 +480,36 @@ the first pattern, followed by one or more `!` exception patterns.
 
 ## FROM
 
-    FROM <image>
+    FROM <image> [AS <name>]
 
 Or
 
-    FROM <image>:<tag>
+    FROM <image>:<tag> [AS <name>]
```
Comment (thaJeztah) [inline docs/reference/builder.md]: this was already like this, but given that tag is optional, should this be?;

   FROM <image>[:<tag>] [AS <name>]

File: docs/reference/builder.md [hunk 4]
```diff
@@ -480,30 +480,36 @@ the first pattern, followed by one or more `!` exception patterns.
...
 
 Or
 
-    FROM <image>@<digest>
+    FROM <image>@<digest> [AS <name>]
```
Comment (thaJeztah) [inline docs/reference/builder.md]: same here for digest

File: docs/reference/builder.md [hunk 5]
```diff
@@ -937,6 +943,13 @@ All new files and directories are created with a UID and GID of 0.
 > If you build using STDIN (`docker build - < somefile`), there is no
 > build context, so `COPY` can't be used.
 
+Optionally `COPY` accepts a flag `--from=[name|index]` that can be used to set
```
Comment (thaJeztah) [inline docs/reference/builder.md]: `--from=<name|index>` (name or index is required when using `--from`)

File: docs/reference/builder.md [hunk 6]
```diff
@@ -937,6 +943,13 @@ All new files and directories are created with a UID and GID of 0.
...
+the source location to a previous build stage(created with `FROM .. AS <name>`
```
Comment (thaJeztah) [inline docs/reference/builder.md]: - add a space before `(`
- "created with _the_ ..." (or remove "command" on the next line)

**Review by thaJeztah — CHANGES_REQUESTED** [review]
some minor nits, but LGTM otherwise

**Comment by tonistiigi** [comment]
@thaJeztah Updated. PTAL

**Review by thaJeztah — APPROVED** [review]
LGTM, thanks!

**Review by vdemeester — APPROVED** [review]
LGTM 🐸

**Pull request merged**